### PR TITLE
_d_assert_fail: Add formatting for Assertion errors

### PR DESCRIPTION
**Pull request by wilzbach**
Requires https://github.com/dlang/dmd/pull/8517

This adds `_d_assert_fail` to `object.d` which allows for nicely formatted and informative error messages on an assertion failure.

```
int a = 1, b = 2;
assert(a == b); // 1 != 2 (and no longer) 
```

Compare at the moment, it would yield sth. like `core.exception.AssertError@onlineapp.d(4): Assertion failure`. But I think we all agree that we can do better.

The printing supported added here is pretty minimalistic as
- this is still behind a version flag
- the program will terminate after this formatting (so saving a few bytes isn't a huge concern)
- Phobos shouldn't be used to reduce the compilation overhead of this feature

tl;dr: yes, the formatting can be improved (and should be before this feature gets activated by default), but this PR adds enough formatting to bootstrap https://github.com/dlang/dmd/pull/8517 as an existing `_d_assert_fail` is required. Also, once this is merged, it will be a lot easier for everyone to improve upon this.

### CODE REVIEW

**Comment by wilzbach** [comment]
Oh and as this depends on the DMD PR, we will have to do a druntime -> dmd -> druntime PR series.
So I disabled the added tests from running by removing `assert_fail` from `TESTS`. Once this is merged and the respective DMD PR, we can actually enable the test. It sucks that the repos are so intertwined.

File: src/object.d [hunk 1]
```diff
@@ -4795,3 +4795,152 @@ TTo[] __ArrayCast(TFrom, TTo)(TFrom[] from) @nogc pure @trusted
...
+private template getPrintfFormat(T)
+{
+    static if (is(T == long))
+    {
+        enum getPrintfFormat = "%lld";
+    }
+    else static if (is(T == ulong))
+    {
+        enum getPrintfFormat = "%llu";
+    }
```
Comment (jacob-carlborg) [inline src/object.d]: According to Clang on macOS:

* `c_long`  -> `%ld` 
* `c_ulong` -> `%lu`
* `size_t` on 64bit-> `%lu` 
* `size_t` on 32bit-> `%zu`

File: src/object.d [hunk 2]
```diff
@@ -4795,3 +4795,152 @@ TTo[] __ArrayCast(TFrom, TTo)(TFrom[] from) @nogc pure @trusted
...
+    import core.stdc.stdio : sprintf;
+    import core.stdc.string : strlen;
+    static if (__traits(isIntegral, V))
+    {
+        enum printfFormat = getPrintfFormat!V;
+        char[20] val;
+        sprintf(val.ptr, printfFormat, v);
+        return val.idup[0 .. strlen(val.ptr)];
+    }
+    else static if (__traits(isFloating, V))
```
Comment (jacob-carlborg) [inline src/object.d]: Why a separate case for floats, why not handle this in `getPrintfFormat`?

File: src/object.d [hunk 3]
```diff
@@ -4795,3 +4795,152 @@ TTo[] __ArrayCast(TFrom, TTo)(TFrom[] from) @nogc pure @trusted
...
+    {
+        char[60] val;
+        sprintf(val.ptr, "%g", v);
+        return val.idup[0 .. strlen(val.ptr)];
+    }
+    else static if (__traits(compiles, { string s = V.init.toString(); }))
+    {
+        return v.toString();
+    }
+    else static if (is(V : string))
```
Comment (jacob-carlborg) [inline src/object.d]: Perhaps any kind of a character array should be printed as a string.

File: src/object.d (hunk 2 above)
Comment (wilzbach) [reply src/object.d]: Because I kept the formatting very naive for floating points numbers and always use `%g` and a very wide buffer for the reasons mentioned above (first version of this, happens before program termination).

File: src/object.d (hunk 3 above)
Comment (wilzbach) [reply src/object.d]: Done.

File: src/object.d (hunk 1 above)
Comment (wilzbach) [reply src/object.d]: We don't need to worry about `size_t` as it will be either `ulong` or `uint`, but I changed the formatting for (u)longs to `%lu` and `%ld`.

Comment (jacob-carlborg) [reply src/object.d]: No, it was correct before. `long long` in C is `%lld` and `unsigned long long is `%llu`. It uses different format specifiers for `long` and `long long`. I was thinking of adding additional cases for `c_long`.

File: src/object.d (hunk 2 above)
Comment (jacob-carlborg) [reply src/object.d]: So you're thinking this will be extended with a more complicated formatting of floats?

File: src/object.d (hunk 1 above)
Comment (jacob-carlborg) [reply src/object.d]: Regarding `size_t`, it's supposed to use `%zu` on 32bit macOS. Not sure what the difference is compared to `%u`.

File: src/object.d [hunk 4]
```diff
@@ -4795,3 +4795,153 @@ TTo[] __ArrayCast(TFrom, TTo)(TFrom[] from) @nogc pure @trusted
     foreach (v; s)
         assert(v == cast(short) 0xabab);
 }
+
+// Yields the appropriate printf format token for a type T
+// Indended to be used by miniFormat
+private template getPrintfFormat(T)
```
Comment (ZombineDev) [inline src/object.d]: Can we move the new code to `core.internal.assert`?

Comment (wilzbach) [reply src/object.d]: Can't name it `assert` as it's a keyword, thus I went for `dassert`.

Comment (ZombineDev) [reply src/object.d]: Sure, thanks!

**Comment by thewilsonator** [comment]
@wilzbach please give this a kick.

**Comment by wilzbach** [comment]
> @wilzbach please give this a kick.

I changed the DMD PR, s.t. it passes without the druntime PR.
So we can first merge the DMD PR and this PR will pass (that's why I re-enabled the tests here).

**Comment by thewilsonator** [comment]
Sounds good.

**Comment by wilzbach** [comment]
> So we can first merge the DMD PR and this PR will pass (that's why I re-enabled the tests here).

Rebased & added a changelog entry.

**Comment by wilzbach** [comment]
Changelog preview:

image

File: changelog/assert.dd [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+struct S { int s; }
+assert(S(0) == S(1)); // ERROR: "S(0) !is S(1)"
+
+int a = 1, b = 2);
+assert(a > b); // ERROR: 1 <= 2
+---
+
+Also if no error message can be generated, it will now fallback to displaying
+the text of the `assert` expression. For example, for this more complicated
+assert expression:
```
Comment (rainers) [inline changelog/assert.dd]: It's the same expression as above.

File: src/core/internal/dassert.d [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        case ">":
+            return "<=";
+        case ">=":
+            return "<";
+        case "is":
+            return "!is";
+        case "!is":
+            return "is";
+        default:
+            return "!=";
```
Comment (rainers) [inline src/core/internal/dassert.d]: I wouldn't default to `!=`, better fall back to something using the original operator name.

Comment (wilzbach) [reply src/core/internal/dassert.d]: Ah actually we whitelist in DMD which tokens are allowed (see https://github.com/dlang/dmd/blob/1ffe034e743f5f7833ce5ed457b1f5eb2012d718/src/dmd/expressionsem.d#L5546), so I think the better course of action is to use an `assert(0)`.

File: changelog/assert.dd (hunk 5 above)
Comment (wilzbach) [reply changelog/assert.dd]: Fixed thanks!

**Comment by wilzbach** [comment]
Any other requirements that this initial implementation needs to fulfill?
It's not intended to be perfect, but at least better than just a non-informational `AssertError`.

**Comment by thewilsonator** [comment]
No, I think this is a good start and we'll see what the feedback is like.
